simplex/engine.go

```
@@ -0,0 +1,313 @@
+		StateSummaryFrontierHandler: common.NewNoOpStateSummaryFrontierHandler(config.Log),
+		AcceptedStateSummaryHandler: common.NewNoOpAcceptedStateSummaryHandler(config.Log),
+		AcceptedFrontierHandler:     common.NewNoOpAcceptedFrontierHandler(config.Log),
+		AcceptedHandler:             common.NewNoOpAcceptedHandler(config.Log),
+		AncestorsHandler:            common.NewNoOpAncestorsHandler(config.Log),
+		PutHandler:                  common.NewNoOpPutHandler(config.Log),
+		QueryHandler:                common.NewNoOpQueryHandler(config.Log),
+		ChitsHandler:                common.NewNoOpChitsHandler(config.Log),
+		Connector:                   config.VM,
+
+		AppHandler:   config.VM,
+		vm:           config.VM,
+		consensusCtx: consensusCtx,
+	}, nil
+}
+
+func (e *Engine) Start(ctx context.Context, _ uint32) error {
+	if e.nonValidator {
+		e.logger.Info("non-validator cannot start simplex engine")
+		return nil
+	}
+
+	e.logger.Info("Starting simplex engine")
+	go e.tick()
+	err := e.epoch.Start()
+	if err != nil {
+		return fmt.Errorf("failed to start simplex epoch: %w", err)
+	}
+
+	e.consensusCtx.State.Set(snow.EngineState{
+		Type:  p2p.EngineType_ENGINE_TYPE_CHAIN,
+		State: snow.NormalOp,
+	})
+	return e.vm.SetState(ctx, snow.NormalOp)
+}
+
+// tick periodically advances the engine's time.
+func (e *Engine) tick() {
+	ticker := time.NewTicker(tickInterval)
+	for {
+		tick := <-ticker.C
+		e.epoch.AdvanceTime(tick)
+	}
+}
+
+func (e *Engine) SimplexMessage(ctx context.Context, nodeID ids.NodeID, msg *p2p.Simplex) error {
+	if e.nonValidator {
+		e.logger.Info("non-validator received simplex message; dropping")
+		return nil
+	}
+
+	simplexMsg, err := e.p2pToSimplexMessage(ctx, msg)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errConvertingMessage, err)
+	}
+
+	return e.epoch.HandleMessage(simplexMsg, nodeID[:])
+}
+
+func (e *Engine) p2pToSimplexMessage(ctx context.Context, msg *p2p.Simplex) (*simplex.Message, error) {
```

we don't want to return any error here. 

Returning an error here will propagate an error up to the handler and it will halt the chain.